### PR TITLE
Sharedp and negative indices

### PR DESCRIPTION
**Pull request by christophejunke**
This pull request is related to the following issues:

- https://github.com/vindarel/cl-str/issues/55
- https://github.com/vindarel/cl-str/issues/56

There is probably still work to do, this is more a draft to talk about the changes

### CODE REVIEW

File: test/test-str.lisp [hunk 1]
```diff
@@ -9,6 +9,16 @@
 (setf prove:*enable-colors* t)
 (plan nil)
 
+(subtest "Index"
```
Comment (christophejunke) [inline test/test-str.lisp:12]: Bad commit

File: test/test-str.lisp [hunk 2]
```diff
@@ -9,6 +9,16 @@
...
+  (is (loop
+         for i from -7 upto 7
+         collect (index "abcde" i))
+      '(0 0 0 1 2 3 4 0 1 2 3 4 5 5 5)
+      "index wraps and saturates"))
+
+(subtest "Slice"
```
Comment (christophejunke) [inline test/test-str.lisp:19]: Needs more tests

File: str.lisp [hunk 3]
```diff
@@ -230,21 +285,15 @@
...
-     (concatenate 'string
-                  (subseq s 0 index)
-                  string/char
-                  (subseq s index)))))
+    ((and s index string/char)
+     ;; insert _ in "abcd" at -1 means "abcd_", not "abc_d"
+     (with-indices (index) (1+ (length s))
+       (concat (slice 0 index s t)
+               (string string/char)
+               (slice index (length s) s t))))
```
Comment (christophejunke) [inline str.lisp]: use nil/t instead of (length s)

File: README.md [hunk 4]
```diff
@@ -271,11 +271,22 @@ Examples:
   (is "" (substring 2 1 "abcd") "start is bigger than end")
 ```
 
+#### slice `(start end s &optional sharedp)`
+
+Same as substring, except the result might share storage with s.
```
Comment (vindarel) [inline README.md:276]: and more importantly, what's a negative `end`.

File: str.lisp [hunk 5]
```diff
@@ -137,27 +151,126 @@
 (defun version ()
   (print +version+))
 
+;; as a dedicated condition to help detect it in tests
+;; (up to 0.18.1 a negative index would mean zero)
```
Comment (vindarel) [inline str.lisp:162]: but not in all functions right? I have to check below.

File: str.lisp [hunk 6]
```diff
@@ -137,27 +151,126 @@
...
+  (let ((length (etypecase string/length
+                  (string (length string/length))
+                  (array-total-size string/length))))
+    (cond
+      ((member index '(nil t)) length)
+      ((= 0 length) 0)
+      (t
+       (locally (declare (type (and fixnum (not (eql 0))) length))
+         (when (< index 0)
+           (warn 'negative-index))
```
Comment (vindarel) [inline str.lisp]: unnecessary to warn for a new function? (it doesn't harm though)

File: str.lisp [hunk 7]
```diff
@@ -137,27 +151,126 @@
...
+;; (up to 0.18.1 a negative index would mean zero)
+(define-condition negative-index (warning) ()
+  (:report
+   "STR possible breaking change: negative index means 'from end' now"))
+
+(declaim (inline index))
+
+(defun index (string/length index)
+  ""
+  (declare (optimize (speed 3))
```
Comment (vindarel) [inline str.lisp]: I don't think it is necessary nor a good thing to hardcode this. Explain?

File: str.lisp [hunk 8]
```diff
@@ -137,27 +151,126 @@
...
+           (warn 'negative-index))
+         (multiple-value-bind (quotient mod) (floor index length)
+           (if (<= -1 quotient 0)
+               ;; wrap when index is between -length and length
+               mod
+               ;; otherwise clamp to either side
+               (if (< quotient 0) 0 length))))))))
+
+(defmacro with-indices ((&rest indices) string &body body)
+  "Ensure INDICES are computed according to INDEX w.r.t. STRING in BODY.
```
Comment (vindarel) [inline str.lisp]: Where does INDEX come from?

File: str.lisp [hunk 9]
```diff
@@ -137,27 +151,126 @@
...
+
+INDICES is a list of either a symbol NAME or a couple (NAME VALUE); if
+only NAME is provided, NAME is expected to be bound in the current
+context to the VALUE associaed with NAME.
+
+Inside BODY, each NAME is bound by LET* to (INDEX STRING VALUE).
+"
+  (let (($string (gensym)))
```
Comment (vindarel) [inline str.lisp]: let's use a real name, we have no need of the strange-looking `$`.

File: str.lisp [hunk 10]
```diff
@@ -137,27 +151,126 @@
...
+    `(let ((,$string ,string))
+       ,(flet ((make-binding (index)
```
Comment (vindarel) [inline str.lisp]: uh what's `make-binding`?

File: str.lisp (hunk 9 above)
Comment (vindarel) [reply str.lisp]: Also you can give an argument to gensym to have a familiar prefix: (gensym "str") => :|str987|.

File: str.lisp [hunk 11]
```diff
@@ -137,27 +151,126 @@
...
+                 (let ((index (if (listp index) index (list index))))
+                   (destructuring-bind (name &optional value) index
+                     (let ((index-expr (or value name)))
+                       `(,name (index ,$string ,index-expr)))))))
+          `(let* ,(mapcar #'make-binding indices)
+             ,@body)))))
+
+(declaim (inline slice%))
+(defun slice% (start end s sharedp)
```
Comment (vindarel) [inline str.lisp]: I am not fond of the `name%` naming either. Could be `do-slice`, `slice-fn`,…

File: str.lisp [hunk 12]
```diff
@@ -137,27 +151,126 @@
...
+  (let ((length (- end start)))
+    (cond
+      ((<= length 0) "")
+      ((not sharedp) (subseq s start end))
+      (t (make-array length
+                     :element-type (array-element-type s)
+                     :displaced-to s
+                     :displaced-index-offset start)))))
+
+(defun slice (start end s &optional (sharedp *sharedp*))
```
Comment (vindarel) [inline str.lisp]: Don't you think `shared` should be a keyword argument? I find them always clearer, we used more of them so far, and who knows, if we need more argument changing the function signature can turn painful.

File: str.lisp [hunk 13]
```diff
@@ -137,27 +151,126 @@
...
+  (when s
+    (with-indices ((start (or start 0)) end) s
```
Comment (vindarel) [inline str.lisp]: mmh… how many lines does the macro save? What would be the equivalent without it?

File: str.lisp [hunk 14]
```diff
@@ -137,27 +151,126 @@
...
+      (slice% start end s sharedp))))
+
+;; internal
+(defun whitespacep (char)
```
Comment (vindarel) [inline str.lisp:245]: oh yes sure +1

File: str.lisp [hunk 15]
```diff
@@ -137,27 +151,126 @@
...
+  (member char *whitespaces*))
+
+;; internal
+(defun trim-left-if (p s)
```
Comment (vindarel) [inline str.lisp:251]: They still accept docstrings ;) (a simple one, at least to inform about the parameters (for which type declarations are ok too ;) ))

Comment (vindarel) [reply str.lisp:251]: Also is `p` for `predicate`? Then the full word is better. I tend to avoid one-letter names. edit: except `s` :p

File: str.lisp [hunk 16]
```diff
@@ -137,27 +151,126 @@
...
+        (end (position-if-not p s :from-end t)))
+    (if (and beg end)
+        (slice beg (1+ end) s)
+        "")))
+
 (defun trim-left (s)
   "Remove whitespaces at the beginning of s. "
   (when s
-    (string-left-trim *whitespaces* s)))
+    (trim-left-if #'whitespacep s)))
```
Comment (vindarel) [inline str.lisp:275]: Clever but please, why? The function was working. What does that solve?

same+2 below.

File: str.lisp [hunk 17]
```diff
@@ -137,27 +151,126 @@
...
-    (string-trim *whitespaces* s)))
+    (trim-if #'whitespacep s)))
 
 (defun collapse-whitespaces (s)
   "Ensure there is only one space character between words.
   Remove newlines."
+  ;; FIXME? use \\s everywhere we need whitespace? trim could be based
+  ;; on regexes too; OR, match against
+  ;; '(:greedy-repetition 1 NIL (:char-class #.*WHITESPACES*))
+  ;; (is *WHITESPACES* supposed to be constant?)
```
Comment (vindarel) [inline str.lisp:296]: `*whitespaces*` is supposed to be a variable.

We can discuss this… on another issue / PR.

File: str.lisp [hunk 18]
```diff
@@ -167,28 +280,20 @@
 (defun join (separator strings)
   "Join all the strings of the list with a separator."
   (let ((separator (replace-all "~" "~~" (string separator))))
-    (format nil
-            (concatenate 'string "~{~a~^" separator "~}")
-            strings)))
+    (format nil (concat "~{~a~^" separator "~}") strings)))
```
Comment (vindarel) [inline str.lisp:306]: So, at first look: this shouldn't part of this PR.

It seems stupid to not use one's own function… however, a little voice tells me why not use the built-in? Using the built-in, pros:

- one less function call (concat calls concatenate)
- a casual reader can learn CL's basics (that's quite important to me).

File: str.lisp (hunk 3 above)
Comment (vindarel) [reply str.lisp]: Here we need a warning!

btw, we could enrich the warning's message, to explain where it comes from, and what is this change.

Comment (vindarel) [reply str.lisp]: > looks doable. With a warning first, and maybe with a key argument to control the behavior? If someone speaks up we can be more conservative.

I would appreciate a keyword to control the behavior. And, I am afraid code of mine would break so I am tempted to say, let the new behavior be optional and activated with the keyword argument.

Could we speak a bit more in the issue?

File: str.lisp [hunk 19]
```diff
@@ -197,9 +302,10 @@
   split at most `limit' - 1 times)."
   ;; cl-ppcre:split doesn't return a null string if the separator appears at the end of s.
   (let* ((limit (or limit (1+ (length s))))
-         (res (cl-ppcre:split (cl-ppcre:quote-meta-chars (string separator)) s :limit limit :start start :end end)))
+         (res (cl-ppcre:split (cl-ppcre:quote-meta-chars (string separator))
+                              s :limit limit :start start :end end :sharedp *sharedp*)))
     (if omit-nulls
-        (remove-if (lambda (it) (empty? it)) res)
+        (delete-if #'empty? res)
```
Comment (vindarel) [inline str.lisp:331]: No. delete-if is destructive, we want to create a new string.

File: str.lisp [hunk 20]
```diff
@@ -250,14 +356,14 @@ It uses `subseq' with differences:
     (let ((end (max (- len (length ellipsis))
                     0)))
       (setf s (concat
-               (subseq s 0 end)
+               (slice% 0 end s t)
```
Comment (vindarel) [inline str.lisp]: Why, what does that solve, is it a breaking change?

File: str.lisp [hunk 21]
```diff
@@ -250,14 +356,14 @@ It uses `subseq' with differences:
...
                ellipsis))))
   s)
 
 (defun words (s &key (limit 0))
   "Return list of words, which were delimited by white space. If the optional limit is 0 (the default), trailing empty strings are removed from the result list (see cl-ppcre)."
   (when s
-      (cl-ppcre:split "\\s+" (trim-left s) :limit limit)))
+    (cl-ppcre:split "\\s+" (trim-left s) :limit limit :sharedp *sharedp*)))
```
Comment (vindarel) [inline str.lisp:389]: add the keyword argument.

File: str.lisp [hunk 22]
```diff
@@ -388,8 +493,26 @@ A simple call to the built-in `search` (which returns the position of the substr
...
+              (eq base (array-displacement array)))
+          array
+          (make-array (length array)
+                      :element-type (array-element-type base)
+                      :displaced-to base
+                      :displaced-index-offset offset)))))
+
 (defun prefix-1 (item1 item2)
-  (subseq item1 0 (or (mismatch item1 item2) (length item1))))
+  (slice 0 (mismatch item1 item2) item1))
```
Comment (vindarel) [inline str.lisp:538]: review to be continued.

**Review by vindarel — CHANGES_REQUESTED** [review]

**Comment by vindarel** [comment]
There are too may changes. Please split the PR.

Some little changes would be ok to merge.

And we must discuss `insert` a bit more.

thanks!

File: str.lisp (hunk 7 above)
Comment (christophejunke) [reply str.lisp]: this is not necessary, might be better to remove declarations

File: str.lisp (hunk 6 above)
Comment (christophejunke) [reply str.lisp]: added a warnp keyword argument thay may eventually be removed later

File: str.lisp (hunk 8 above)
Comment (christophejunke) [reply str.lisp]: Function INDEX, updating comment

File: str.lisp (hunk 10 above)
Comment (christophejunke) [reply str.lisp]: renaming it as `let-binding` since it produces a binding for let for each symbol

File: str.lisp (hunk 9 above)
Comment (christophejunke) [reply str.lisp]: done

File: str.lisp (hunk 11 above)
Comment (christophejunke) [reply str.lisp]: going for `unsafe-slice` if that's ok

File: str.lisp (hunk 12 above)
Comment (christophejunke) [reply str.lisp]: I agree keyword arguments are generally a better option

File: str.lisp (hunk 13 above)
Comment (christophejunke) [reply str.lisp]: This is less about saving lines and more about having a descriptive name for what is being computed. Whenever there is code that relies on a string index we can wrap the existing code with this.
Currently this is equivalent to:

    (let* ((start (index s (or start 0)))
             (end (index s end)))
        ....)

But this may evolve later to do things better, without having to update the call (e.g. add declarations)

File: str.lisp (hunk 14 above)
Comment (christophejunke) [reply str.lisp:245]: I should probably export it and have a whitespace? alias

File: str.lisp (hunk 3 above)
Comment (christophejunke) [reply str.lisp]: Yes, I am thinking about adding some `if-out-of-bounds` keyword argument, we can talk more about it in the issue, if you have questions or preferences.
